File: bindings/go/scoping.go
package forge

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
	"unicode/utf8"
)

const maxScopeComponentBytes = 255

func ScopeKVKey(application, tenant, user, resource string) (string, error) {
	return renderScope("kv", 383, application, tenant, user, resource)
}

func ScopeBlobKey(application, tenant, user, resource string) (string, error) {
	return renderScope("blob", 895, application, tenant, user, resource)
}

func ScopeRateLimitSubject(application, tenant, user, resource string) (string, error) {
	return renderScope("rate", 383, application, tenant, user, resource)
}

func ScopeTopic(application, tenant, user, resource string) (string, error) {
	return renderScope("topic", 383, application, tenant, user, resource)
}

func renderScope(kind string, budget int, application, tenant, user, resource string) (string, error) {
	parts := []struct{ label, value string }{
		{"application", application},
		{"tenant", tenant},
		{"user", user},
		{"resource", resource},
	}
	for _, part := range parts {
		if err := validateScopeComponent(part.label, part.value); err != nil {
			return "", err
		}
	}
	value := fmt.Sprintf("v1|%s|%d:%s%d:%s%d:%s%d:%s", kind, len(application), application, len(tenant), tenant, len(user), user, len(resource), resource)
	if len(value) > budget {
		return "", forgeError(CodeLimit, "scope."+kind, "scoped "+kind+" name exceeds its backend-safe length")
	}
	return value, nil
}

type ParsedScope struct {
	Kind, Application, Tenant, User, Resource string
}

func ParseScopedName(value string) (ParsedScope, error) {
	if !strings.HasPrefix(value, "v1|") {
		return ParsedScope{}, forgeError(CodeInvalid, "scope.parse", "scoped name must use v1")
	}
	rest := strings.TrimPrefix(value, "v1|")
	separator := strings.IndexByte(rest, '|')
	if separator < 0 {
		return ParsedScope{}, forgeError(CodeInvalid, "scope.parse", "scoped name is malformed")
	}
	kind, encoded := rest[:separator], rest[separator+1:]
	budget := 383
	if kind == "blob" {
		budget = 895
	} else if kind != "kv" && kind != "rate" && kind != "topic" {
		return ParsedScope{}, forgeError(CodeInvalid, "scope.parse", "scoped name kind is unknown")
	}
	labels := []string{"application", "tenant", "user", "resource"}
	parts := make([]string, 0, len(labels))
	for _, label := range labels {
		colon := strings.IndexByte(encoded, ':')
		if colon < 1 {
			return ParsedScope{}, forgeError(CodeInvalid, "scope.parse", "scoped name is malformed")
		}
		lengthText := encoded[:colon]
		if strings.IndexFunc(lengthText, func(value rune) bool { return value < '0' || value > '9' }) >= 0 {
			return ParsedScope{}, forgeError(CodeInvalid, "scope.parse", "scoped name length is malformed")
		}
		length, err := strconv.Atoi(lengthText)
		if err != nil || colon+1+length > len(encoded) {
			return ParsedScope{}, forgeError(CodeInvalid, "scope.parse", "scoped name component length is invalid")
		}
		part := encoded[colon+1 : colon+1+length]
		if err := validateScopeComponent(label, part); err != nil {
			return ParsedScope{}, err
		}
		parts = append(parts, part)
		encoded = encoded[colon+1+length:]
	}
	if encoded != "" || len(value) > budget {
		return ParsedScope{}, forgeError(CodeInvalid, "scope.parse", "scoped name has trailing or oversized data")
	}
	return ParsedScope{Kind: kind, Application: parts[0], Tenant: parts[1], User: parts[2], Resource: parts[3]}, nil
}

func validateScopeComponent(label, component string) error {
	if len(component) == 0 || len(component) > maxScopeComponentBytes {
		return forgeError(CodeInvalid, "scope", "scope "+label+" must contain 1 to 255 bytes")
	}
	if !utf8.ValidString(component) {
		return forgeError(CodeInvalid, "scope", "scope "+label+" must be valid UTF-8")
	}
	if strings.IndexFunc(component, unicode.IsControl) >= 0 {
		return forgeError(CodeInvalid, "scope", "scope "+label+" must not contain control characters")
	}
	return nil
}

File: src/scoping.rs
use crate::{ForgeError, Result};

const MAX_COMPONENT_BYTES: usize = 255;
const KV_LOGICAL_BUDGET: usize = 383;
const BLOB_LOGICAL_BUDGET: usize = 895;

fn validate_component(label: &str, component: &str) -> Result<()> {
    if component.is_empty() || component.len() > MAX_COMPONENT_BYTES {
        return Err(ForgeError::invalid(format!(
            "scope {label} must contain 1 to 255 bytes"
        )));
    }
    if component.chars().any(char::is_control) {
        return Err(ForgeError::invalid(format!(
            "scope {label} must not contain control characters"
        )));
    }
    Ok(())
}

fn render_scoped_name(
    kind: &str,
    budget: usize,
    application: &str,
    tenant: &str,
    user: &str,
    resource: &str,
) -> Result<String> {
    for (label, component) in [
        ("application", application),
        ("tenant", tenant),
        ("user", user),
        ("resource", resource),
    ] {
        validate_component(label, component)?;
    }
    let value = format!(
        "v1|{kind}|{}:{}{}:{}{}:{}{}:{}",
        application.len(),
        application,
        tenant.len(),
        tenant,
        user.len(),
        user,
        resource.len(),
        resource
    );
    if value.len() > budget {
        return Err(ForgeError::limit(format!(
            "scoped {kind} name exceeds its backend-safe length"
        )));
    }
    Ok(value)
}

pub fn scope_kv_key(application: &str, tenant: &str, user: &str, resource: &str) -> Result<String> {
    render_scoped_name("kv", KV_LOGICAL_BUDGET, application, tenant, user, resource)
}

pub fn scope_blob_key(
    application: &str,
    tenant: &str,
    user: &str,
    resource: &str,
) -> Result<String> {
    render_scoped_name(
        "blob",
        BLOB_LOGICAL_BUDGET,
        application,
        tenant,
        user,
        resource,
    )
}

pub fn scope_rate_limit_subject(
    application: &str,
    tenant: &str,
    user: &str,
    resource: &str,
) -> Result<String> {
    render_scoped_name(
        "rate",
        KV_LOGICAL_BUDGET,
        application,
        tenant,
        user,
        resource,
    )
}

pub fn scope_topic(application: &str, tenant: &str, user: &str, resource: &str) -> Result<String> {
    render_scoped_name(
        "topic",
        KV_LOGICAL_BUDGET,
        application,
        tenant,
        user,
        resource,
    )
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedScope {
    pub kind: String,
    pub application: String,
    pub tenant: String,
    pub user: String,
    pub resource: String,
}

pub fn parse_scoped_name(value: &str) -> Result<ParsedScope> {
    let rest = value
        .strip_prefix("v1|")
        .ok_or_else(|| ForgeError::invalid("scoped name must use v1"))?;
    let (kind, mut encoded) = rest
        .split_once('|')
        .ok_or_else(|| ForgeError::invalid("scoped name is malformed"))?;
    let budget = match kind {
        "blob" => BLOB_LOGICAL_BUDGET,
        "kv" | "rate" | "topic" => KV_LOGICAL_BUDGET,
        _ => return Err(ForgeError::invalid("scoped name kind is unknown")),
    };
    if value.len() > budget {
        return Err(ForgeError::limit(format!(
            "scoped {kind} name exceeds its backend-safe length"
        )));
    }
    let mut components = Vec::with_capacity(4);
    for label in ["application", "tenant", "user", "resource"] {
        let (length, tail) = encoded
            .split_once(':')
            .ok_or_else(|| ForgeError::invalid("scoped name is malformed"))?;
        if length.is_empty() || !length.bytes().all(|byte| byte.is_ascii_digit()) {
            return Err(ForgeError::invalid("scoped name length is malformed"));
        }
        let length: usize = length
            .parse()
            .map_err(|_| ForgeError::invalid("scoped name length is malformed"))?;
        let component = tail
            .as_bytes()
            .get(..length)
            .and_then(|part| std::str::from_utf8(part).ok())
            .ok_or_else(|| ForgeError::invalid("scoped name component length is invalid"))?;
        validate_component(label, component)?;
        components.push(component.to_string());
        encoded = tail
            .get(length..)
            .ok_or_else(|| ForgeError::invalid("scoped name component length is invalid"))?;
    }
    if !encoded.is_empty() {
        return Err(ForgeError::invalid("scoped name has trailing data"));
    }
    let [application, tenant, user, resource]: [String; 4] = components
        .try_into()
        .map_err(|_| ForgeError::invalid("scoped name has the wrong component count"))?;
    Ok(ParsedScope {
        kind: kind.to_string(),
        application,
        tenant,
        user,
        resource,
    })
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
    use super::*;

    #[test]
    fn scoped_names_are_reversible_and_primitive_specific() {
        let vectors: serde_json::Value =
            serde_json::from_str(include_str!("../contract/scope-vectors.json")).unwrap();
        let valid = vectors.get("valid").unwrap();
        let component = |name: &str| valid.get(name).and_then(serde_json::Value::as_str).unwrap();
        let args = (
            component("application"),
            component("tenant"),
            component("user"),
            component("resource"),
        );
        let kv = scope_kv_key(args.0, args.1, args.2, args.3).unwrap();
        assert_eq!(kv, component("kv"));
        assert_eq!(
            scope_blob_key(args.0, args.1, args.2, args.3).unwrap(),
            component("blob")
        );
        assert_eq!(
            scope_rate_limit_subject(args.0, args.1, args.2, args.3).unwrap(),
            component("rate")
        );
        assert_eq!(
            scope_topic(args.0, args.1, args.2, args.3).unwrap(),
            component("topic")
        );
        assert_eq!(
            parse_scoped_name(&kv).unwrap(),
            ParsedScope {
                kind: "kv".into(),
                application: args.0.into(),
                tenant: args.1.into(),
                user: args.2.into(),
                resource: args.3.into(),
            }
        );
    }

    #[test]
    fn invalid_names_fail_before_backend_io() {
        assert!(scope_kv_key("", "t", "u", "r").is_err());
        assert!(scope_topic("app", "t", "u", "a\nb").is_err());
        assert!(parse_scoped_name("v1|kv|+3:app1:t1:u1:r").is_err());
        let long = "x".repeat(100);
        assert!(scope_kv_key(&long, &long, &long, &long).is_err());
        assert!(scope_blob_key(&long, &long, &long, &long).is_ok());
    }
}
